service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java

```diff
@@ -2041,49 +2041,71 @@ public void saveAdminCommentToOrder(AdminCommentDto adminCommentDto, String uuid
             () -> new OrderNotFoundException(ORDER_WITH_CURRENT_ID_DOES_NOT_EXIST + orderId));
-        if (ecoNumberDto != null) {
-            StringBuilder collectedValue = new StringBuilder();
-            for (int i = 0; i < ecoNumberDto.size(); i++) {
-                EcoNumberDto ecoNumber = ecoNumberDto.get(i);
-                String oldNumber = orderRepository.findEcoNumberFromShop(ecoNumber.getOldEcoNumber(), orderId);
-                if (oldNumber != null) {
-                    orderRepository.setOrderAdditionalNumber(ecoNumber.getNewEcoNumber(), oldNumber, orderId);
-                    collectedValue.append(
-                        collectInfoAboutEcoNumberEventHistory(i, oldNumber, ecoNumber.getNewEcoNumber()));
-                }
+        Set<String> oldEcoNumbers = Set.copyOf(order.getAdditionalOrders());
+        Set<String> newEcoNumbers = ecoNumberDto.getEcoNumber();
+
+        int countOfNewNumbers = 0;
+        int countOfDeleted = 0;
+
+        StringBuilder collectedNewValue = new StringBuilder();
+        StringBuilder collectedDeletedValue = new StringBuilder();
+        for (String newNumber : newEcoNumbers) {
+            if (!oldEcoNumbers.contains(newNumber)
+                && !newNumber.equals("")) {
+                order.getAdditionalOrders().add(newNumber);
+                collectedNewValue.append(collectInfoAboutAddNewEcoNumber(countOfNewNumbers, newNumber));
+                countOfNewNumbers++;
             }
-            if (collectedValue.length() > 1) {
-                eventService.save(collectedValue.toString(),
-                    currentUser.getRecipientName() + "  " + currentUser.getRecipientSurname(), order);
+        }
+        for (String oldNumber : oldEcoNumbers) {
+            if (!newEcoNumbers.contains(oldNumber)) {
+                order.getAdditionalOrders().remove(oldNumber);
+                collectedDeletedValue.append(collectInfoAboutDeletedEcoNumber(countOfDeleted, oldNumber));
+                countOfDeleted++;
             }
         }
+
+        orderRepository.save(order);
+        eventService.save(allHistoryChanges(collectedNewValue, collectedDeletedValue),
+            currentUser.getRecipientName() + "  " + currentUser.getRecipientSurname(), order);
     }
 
-    /**
-     * This is method which collects info about eco number for event history.
-     *
-     * @author Yuriy Bahlay.
-     */
-    private String collectInfoAboutEcoNumberEventHistory(int i, String oldNumber, String newEcoNumber) {
+    private String allHistoryChanges(
+        StringBuilder collectedNewValue,
+        StringBuilder collectedDeletedValue) {
+        StringBuilder finalHistory = new StringBuilder();
+        if (!collectedNewValue.toString().isEmpty()) {
+            finalHistory.append(collectedNewValue);
+        }
+        if (!collectedDeletedValue.toString().isEmpty()) {
+            finalHistory.append(collectedDeletedValue);
+        }
+        return finalHistory.toString();
+    }
```

Comment:
Can we move all the history functionality to eventService instead of keeping it in UBSManagementServiceImpl? This service needs refactoring.

Reply:
Yeah, that's a good idea